### PR TITLE
feat(sdk): aws.Function.context() for lambda functions

### PR DESCRIPTION
**Pull request by Chriscbr**
Closes https://github.com/winglang/wing/issues/6412

## Checklist

- [x] Title matches [Winglang's style guide](https://www.winglang.io/contributing/start-here/pull_requests#how-are-pull-request-titles-formatted)
- [x] Description explains motivation and solution
- [x] Tests added (always)
- [ ] Docs updated (only required for features)
- [ ] Added `pr/e2e-full` label if this feature requires end-to-end testing

*By submitting this pull request, I confirm that my contribution is made under the terms of the [Wing Cloud Contribution License](https://github.com/winglang/wing/blob/main/CONTRIBUTION_LICENSE.md)*.

### CODE REVIEW

File: examples/tests/sdk_tests/function/aws-function.test.w [hunk 1]
```diff
@@ -35,4 +36,14 @@ test "validates the AWS Function" {
     // If the test is not on AWS, it should not fail, so I am returning true.
     assert(true);
   }
-}
\ No newline at end of file
+}
+
+if target == "tf-aws" {
+  test "can access lambda context" {
+    let ctx = aws.context();
```
Comment (eladb) [inline examples/tests/sdk_tests/function/aws-function.test.w:43]:
```suggestion
    let ctx = aws.Function.context();
```

I think this to be semantically tied to `aws.Function`. 

It should also return an optional and resolve to `nil` on non-AWS functions and in the simulator.

**Review by eladb — CHANGES_REQUESTED** [review]

File: libs/wingc/src/type_check/jsii_importer.rs [hunk 2]
```diff
@@ -583,9 +583,6 @@ impl<'a> JsiiImporter<'a> {
 		if let Some(properties) = jsii_interface.properties() {
 			for p in properties {
 				debug!("Found property {} with type {:?}", p.name.green(), p.type_);
-				if member_phase == Phase::Inflight {
-					todo!("No support for inflight properties yet");
-				}
```
Comment (Chriscbr) [inline libs/wingc/src/type_check/jsii_importer.rs:588]: FYI @yoav-steinberg

File: examples/tests/sdk_tests/function/aws-function.test.w (hunk 1 above)
Comment (Chriscbr) [reply examples/tests/sdk_tests/function/aws-function.test.w]: Done

**Review by MarkMcCulloh — APPROVED** [review]

File: libs/awscdk/src/function.ts [hunk 3]
```diff
@@ -214,4 +214,25 @@ export class Function
...
+   */
+  protected _getCodeLines(handler: cloud.IFunctionHandler): string[] {
+    const inflightClient = handler._toInflight();
+    const lines = new Array<string>();
+    const client = "$handler";
+
+    lines.push('"use strict";');
+    lines.push(`var ${client} = undefined;`);
+    lines.push("exports.handler = async function(event, context) {");
+    lines.push("  globalThis.$awsLambdaContext = context;");
```
Comment (eladb) [inline libs/awscdk/src/function.ts:229]: I am concerned that multiple concurrent invocations will corrupt this global state.

Can we use https://nodejs.org/api/async_context.html?

File: libs/wingsdk/src/target-tf-aws/function.ts [hunk 4]
```diff
@@ -425,7 +425,8 @@ export class Function extends cloud.Function implements IAwsFunction {
 
     lines.push('"use strict";');
     lines.push(`var ${client} = undefined;`);
-    lines.push("exports.handler = async function(event) {");
+    lines.push("exports.handler = async function(event, context) {");
```
Comment (eladb) [inline libs/wingsdk/src/target-tf-aws/function.ts:428]: We should reuse this code across the two aws platforms

**Review by eladb — CHANGES_REQUESTED** [review]

File: libs/awscdk/src/function.ts (hunk 3 above)
Comment (Chriscbr) [reply libs/awscdk/src/function.ts:229]: @eladb can you explain? Concurrent invocations would have to be running on separate AWS Lambda instances, wouldn't they?

Comment (eladb) [reply libs/awscdk/src/function.ts:229]: Yes, that might be guaranteed by the runtime but not enforced by the programming model. There is a reason why `ctx` is an argument to the handler and not a global.

I would say that the "correct" solution is to use the local storage, and perhaps that's something we want to formalize in wing as well.

but if it's too complex, I would at least add some defensive coding which nullifies the global at the end of the run (in a finally block) and verified that it is null at the beginning.

Comment (Chriscbr) [reply libs/awscdk/src/function.ts:229]: IMO if AWS is providing this as a contract, then it should be alright to optimize our private implementations around it -- the only issue would be if we introduce a bug where we start using these AWS Lambda wrappers on platforms that aren't AWS Lambda.

I like the idea of adding some error checking so at least we can catch if this assumption is broken 👍

**Review by eladb — APPROVED** [review]

**Pull request merged**